File: C++/367. Valid Perfect Square/binary-search.cpp
// Time : O(logn)
// Space : O(1)

class Solution {
private:
    // predicate : mid * mid >= num
    bool p(int num, int mid) {
        return mid >= num/mid;
    }
public:
    bool isPerfectSquare(int num) {
        int lo = 1;
        int hi = num;
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (p(num, mid)) {
                hi = mid;
            }
            else {
                lo = mid + 1;
            }
        }
        
        if (lo * lo == num) return true;
        else return false;
    }
};

File: 367. Valid Perfect Square/binary-search.py
class Solution(object):
    def isPerfectSquare(self, num):
        """
        :type num: int
        :rtype: bool
        find a first x for which x * x >= num is true.
        """
        lo = 1
        hi = num
        while lo < hi:
            mid = lo + (hi - lo) // 2
            if mid * mid >= num:
                hi = mid
            else:
                lo = mid + 1
        print(lo)
        return lo * lo == num